### ISSUE TITLE
Use Gradle configuration cache

### ISSUE THREAD

**Issue by liblit** [wala/WALA#1206]
WALA's Gradle build system is complex, to put it mildly. We should try enabling [Gradle's configuration cache](https://docs.gradle.org/current/userguide/configuration_cache.html). If we can get that working, incremental builds should be significantly faster.

**Comment by liblit** [comment]
Assigning to me. I'll take a first swing at this, see how far I get, and document both progress and blockers here.

**Comment by liblit** [comment]
The [`gradle-configuration-cache`](https://github.com/wala/WALA/tree/gradle-configuration-cache) branch enables the Gradle configuration cache. [Builds on this branch currently fail](https://github.com/wala/WALA/actions/runs/3933570033), because we still have some work to do. We already handled problems with the Google Java format plugin in #1209, so that's progress. Errors reported on [the current `HEAD` of the `gradle-configuration-cache` branch](https://github.com/wala/WALA/commit/6d2b0d4985fd550dd5a36b65ae61d695a9bc5415) include:

> 63 problems were found storing the configuration cache, 9 of which seem unique.
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'java.io.FileDescriptor', a subtype of 'java.io.FileDescriptor', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'java.io.PrintStream', a subtype of 'java.io.OutputStream', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'java.io.RandomAccessFile', a subtype of 'java.io.RandomAccessFile', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.internal.classloader.VisitableURLClassLoader', a subtype of 'java.lang.ClassLoader', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.internal.component.external.model.PreferJavaRuntimeVariant', a subtype of 'org.gradle.api.attributes.AttributesSchema', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.internal.concurrent.DefaultExecutorFactory$TrackedManagedExecutor', a subtype of 'java.util.concurrent.Executor', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.internal.concurrent.DefaultExecutorFactory$TrackedScheduledManagedExecutor', a subtype of 'java.util.concurrent.Executor', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.launcher.daemon.server.exec.LogToClient$AsynchronousLogDispatcher', a subtype of 'java.lang.Thread', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:shellCheck` of type `org.gradle.api.tasks.Exec`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types

This is just a summary of a more complete HTML report that GitHub will not let me attach here. The complete report confirms what we see in this summary, which is that the reported problems arise from two tasks:

1. the `:com.ibm.wala.cast:cast:compileDebugCpp` task of type `org.gradle.language.cpp.tasks.CppCompile` and
2. the `:shellCheck` of type `org.gradle.api.tasks.Exec`

Unfortunately, `CppCompile` does not yet support the configuration cache. We will be blocked on gradle/gradle#13484 and gradle/gradle#13485 for support via the `cpp-application` and `cpp-library` plugins, respectively. There's nothing we can do about these short of dropping WALA's native-code integrations.

We can, however, look more closely at the `:shellCheck` task. `Exec` tasks (from the `base` plugin) should already be supported in general, per gradle/gradle#13455. So any problems with this task are probably our doing. If we fix that task, then at least we'll be ready whenever Gradle's own C++ plugins get caught up.

**Comment by liblit** [comment]
We can make more progress by disabling tasks that relate to native code, as in:

```
./gradlew javadoc build -PexcludeSlowTests --continue --exclude-task=:com.ibm.wala.cast:{smoke_main:checkSmokeMain,{cast,smoke_main,xlator_test}:compile{Debug,Release}Cpp}
```

This build fails with an exciting new set of problems:

> 7 problems were found storing the configuration cache.
> - Task `:com.ibm.wala.cast:test` of type `org.gradle.api.tasks.testing.Test`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.core:extractBcel` of type `org.gradle.api.DefaultTask`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.core:extractBcel` of type `org.gradle.api.DefaultTask`: cannot serialize object of type 'com.ibm.wala.gradle.VerifiedDownload', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:task_access
> - Task `:com.ibm.wala.core:extractKawa` of type `org.gradle.api.DefaultTask`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.core:extractKawa` of type `org.gradle.api.DefaultTask`: cannot serialize object of type 'com.ibm.wala.gradle.VerifiedDownload', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:task_access
> - Task `:com.ibm.wala.core:unpackKawaChess` of type `org.gradle.api.DefaultTask`: cannot serialize Gradle script object references as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.core:unpackKawaChess` of type `org.gradle.api.DefaultTask`: cannot serialize object of type 'com.ibm.wala.gradle.VerifiedDownload', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.

So it looks like we need to [dive into our custom checksum-checking download task class (`VerifiedDownload`)](https://github.com/michel-kraemer/gradle-download-task/issues/284) and various *ad hoc* tasks that we use to unpack or extract material from those downloads.

**Comment by msridhar** [comment]
> Unfortunately, `CppCompile` does not yet support the configuration cache. We will be blocked on [gradle/gradle#13484](https://github.com/gradle/gradle/issues/13484) and [gradle/gradle#13485](https://github.com/gradle/gradle/issues/13485) for support via the `cpp-application` and `cpp-library` plugins, respectively. There's nothing we can do about these short of dropping WALA's native-code integrations.

I'm not even sure I know what WALA's native code does.  @juliandolby is this native code still used / useful?  My concern here is I'm not sure that updating the `cpp-application` and `cpp-library` plugins to be configuration-cache compliant is a priority, and it'd be nice for us to turn on configuration cache (configuration times have moderately increased since we switched to Kotlin build scripts).  If our native code is not being used actively I might be in favor of dropping the native-code integrations.  Of course this decision can wait until we've fixed other configuration-cache-related issues.

**Comment by liblit** [comment]
It's worth noting that the configuration cache need not be an all-or-nothing choice. We could leave it disabled by default, and let individual users add `--configuration-cache` to their Gradle command lines when they want to enable it. If the build targets do not involve any unsupported tasks, then the configuration cache will work just fine. For example, you could run Gradle with a set of target tasks that don't require building any of this native stuff. Of course it would be nicer to not have to think about that, but it's an option we can consider.

That being said, [our custom `VerifiedDownload` task is also not yet compatible with the Gradle configuration cache](https://github.com/michel-kraemer/gradle-download-task/issues/284). So those downloading tasks (and any tests that depend on them) would also need to be avoided, which might not be practical. Perhaps we should wait and see how things play out with michel-kraemer/gradle-download-task#284 before deciding what to do about the native code problems.

**Comment by liblit** [comment]
[The current `gradle-configuration-cache` branch head](https://github.com/wala/WALA/tree/5c4e6395852c430ee884cac4bd5ca385e0aa67f1) includes an IntelliJ IDEA run configuration that tries to build `aggregatedJavadocs`, `build`, `publishAllPublicationsToFakeRemoteRepository`, and `shellcheck`. However, this run configuration also disables all native C++ compilation tasks, as these are known to be incompatible with the Gradle configuration cache, per gradle/gradle#13484 and gradle/gradle#13485. This run configuration also disables all `Sign` tasks, since [these will apparently not be compatible with the Gradle configuration cache until Gradle 8.1](https://github.com/gradle/gradle/issues/13470).

Building using this run configuration yields the following error messages:

> 3 problems were found storing the configuration cache.
> - Task `:com.ibm.wala.core:extractBcel` of type `org.gradle.api.DefaultTask`: cannot serialize object of type 'com.ibm.wala.gradle.VerifiedDownload', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:task_access
> - Task `:com.ibm.wala.core:extractKawa` of type `org.gradle.api.DefaultTask`: cannot serialize object of type 'com.ibm.wala.gradle.VerifiedDownload', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:task_access
> - Task `:com.ibm.wala.core:unpackKawaChess` of type `org.gradle.api.DefaultTask`: cannot serialize object of type 'com.ibm.wala.gradle.VerifiedDownload', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:task_access

These errors all involve [our custom `VerifiedDownload` task class](https://github.com/wala/WALA/blob/master/buildSrc/src/main/kotlin/com/ibm/wala/gradle/VerifiedDownload.kt), which in turn builds upon [the third-party `gradle-download-task` plugin](https://github.com/michel-kraemer/gradle-download-task/issues/284). I already have [a question out to that project asking for guidance as to how we can use that plugin's features in a configuration-cache-safe manner](https://github.com/michel-kraemer/gradle-download-task/issues/284). There's probably nothing more for us to now but wait for a reply.

**Comment by liblit** [comment]
We've pushed this goal forward as far as we can for now. [Remaining open items](https://github.com/orgs/wala/projects/1/views/2) include:

* configuration cache support for standard Gradle C++ tasks
  * gradle/gradle#13484
  * gradle/gradle#13485
  * neither of these tasks is assigned to anyone right now
* configuration cache support for standard Gradle artifact-signing tasks
  * gradle/gradle#13470
  * already implemented
  * expected to be available in Gradle 8.1 RC1

---

Are you eager to try using the configuration cache in spite of these limitations? You can! Run `./gradle` with the `--configuration-cache` flag. If the tasks that you ask Gradle to build do not depend on C++ compilation or artifact signing, then the cache should work as intended. If you ask Gradle to do something that requires C++ compilation or artifact signing, then you'll see warnings advising you that the configuration cache could not be used. The build can still succeed, but you won't gain the performance benefits of configuration caching.

For example:

* `./gradlew --configuration-cache spotlessApply` will use the configuration cache.
* `./gradlew --configuration-cache :com.ibm.wala.core:test` will use the configuration cache.
* `./gradlew --configuration-cache build` will not use the configuration cache, because a complete build requires compiling some C++ components.
* `./gradlew --configuration-cache test` will not use the configuration cache, because some tests require compiling some C++ components.

**Comment by msridhar** [comment]
This is fantastic, thanks for all this work @liblit!  I think I may have found another problem.  I tried running `./gradlew --configuration-cache compileJava compileTestJava`, which I expected to succeed.  Instead, I got a failure:
```
* What went wrong:
Configuration cache problems found in this build.

1 problem was found storing the configuration cache.
- Build file 'build.gradle.kts': external process started 'uname -a'
  See https://docs.gradle.org/7.6/userguide/configuration_cache.html#config_cache:requirements:external_processes
```

In the report file, here is some of the failure stack:
```
org.gradle.api.InvalidUserCodeException: Starting an external process 'uname -a' during configuration time is unsupported.
	at org.gradle.configurationcache.initialization.DefaultConfigurationCacheProblemsListener.onExternalProcessStarted(ConfigurationCacheProblemsListener.kt:84)
	at org.gradle.configurationcache.InstrumentedInputAccessListener.externalProcessStarted(InstrumentedInputAccessListener.kt:110)
	at org.gradle.internal.classpath.Instrumented.externalProcessStarted(Instrumented.java:394)
	at org.gradle.internal.classpath.Instrumented.externalProcessStarted(Instrumented.java:398)
	at org.gradle.internal.classpath.Instrumented.exec(Instrumented.java:259)
	at com.diffplug.common.swt.os.OS.exec(OS.java:156)
	at com.diffplug.common.swt.os.OS.calculateNative(OS.java:133)
	at com.diffplug.common.swt.os.OS.detectPlatform(OS.java:113)
	at com.diffplug.common.swt.os.OS.detectPlatform(OS.java:120)
	at com.diffplug.common.swt.os.OS.getNative(OS.java:100)
	at com.diffplug.gradle.FileMisc.<clinit>(FileMisc.java:426)
	at com.diffplug.gradle.eclipse.MavenCentralMapping.createVersionMap(MavenCentralMapping.java:138)
	at com.diffplug.gradle.eclipse.MavenCentralMapping.groupIdArtifactIdToVersion(MavenCentralMapping.java:124)
	at com.diffplug.gradle.eclipse.MavenCentralExtension$ReleaseConfigurer.<init>(MavenCentralExtension.java:56)
	at com.diffplug.gradle.eclipse.MavenCentralExtension.release(MavenCentralExtension.java:47)
	at com.diffplug.gradle.eclipse.MavenCentralExtension.release(MavenCentralExtension.java:43)
```

This is on my Mac with Java 11.  It looks like it may be an issue in the `com.diffplug.eclipse.mavencentral` plugin.  Just FYI @liblit, if you have time to take a look, maybe we can get another upstream issue fixed :-)

**Comment by msridhar** [comment]
Also, configuration cache is awesome!  Works great for me with `compileJava`, really speeds things up a lot.  Thanks again for your work on this, @liblit!

**Comment by liblit** [comment]
> I think I may have found another problem.

Can I ask you to file that as its own issue, assigned to me? I expect we will have several such glitches, and I don’t want #1206 to become an unwieldy collection of all of them. Thanks!

**Comment by liblit** [comment]
> Also, configuration cache is awesome! Works great for me with `compileJava`, really speeds things up a lot.

Yay! I’m glad you’re finding even the current, partial support useful.

**Comment by liblit** [comment]
Should we turn on configuration caching by default?

* **Pro**
  * speeds up builds, and especially rebuilds, that do not involve compiling native code or signing build artifacts
  * makes any configuration-cache regressions more likely to be discovered earlier
* **Con**
  * produces many warnings in builds that involve compiling native code or signing build artifacts, including the standard `build` target
  * is officially an [incubating](https://docs.gradle.org/current/userguide/feature_lifecycle.html#sec:incubating_state) feature, still subject to change

The "many warnings" I refer to above would typically look like this:

> 2539 problems were found storing the configuration cache, 139 of which seem unique.
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'com.esotericsoftware.kryo.io.Output', a subtype of 'java.io.OutputStream', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'java.io.FileDescriptor', a subtype of 'java.io.FileDescriptor', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'java.io.PrintStream', a subtype of 'java.io.OutputStream', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'java.io.RandomAccessFile', a subtype of 'java.io.RandomAccessFile', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'java.lang.Thread', a subtype of 'java.lang.Thread', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'jdk.internal.loader.ClassLoaders$AppClassLoader', a subtype of 'java.lang.ClassLoader', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'jdk.internal.loader.ClassLoaders$PlatformClassLoader', a subtype of 'java.lang.ClassLoader', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.cache.internal.btree.ByteInput$ResettableBufferedInputStream', a subtype of 'java.io.InputStream', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.cache.internal.btree.ByteOutput$ResettableBufferedOutputStream', a subtype of 'java.io.OutputStream', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.cache.internal.btree.KeyHasher$MessageDigestStream', a subtype of 'java.io.OutputStream', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.initialization.MixInLegacyTypesClassLoader', a subtype of 'java.lang.ClassLoader', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.internal.classloader.FilteringClassLoader', a subtype of 'java.lang.ClassLoader', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.internal.classloader.VisitableURLClassLoader', a subtype of 'java.lang.ClassLoader', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.internal.component.external.model.PreferJavaRuntimeVariant', a subtype of 'org.gradle.api.attributes.AttributesSchema', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> - Task `:com.ibm.wala.cast:cast:compileDebugCpp` of type `org.gradle.language.cpp.tasks.CppCompile`: cannot serialize object of type 'org.gradle.internal.concurrent.DefaultExecutorFactory$TrackedManagedExecutor', a subtype of 'java.util.concurrent.Executor', as these are not supported with the configuration cache.
>   See https://docs.gradle.org/8.0.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
> plus 124 more problems. Please see the report for details.
> 
> See the complete report at file:///.../build/reports/configuration-cache/1fl2mgpfb2zn94ohduj0s5soe/1w89aah7jnkxb0zhwrqo0ryvi/configuration-cache-report.html
> 
> BUILD SUCCESSFUL in 2s
> 321 actionable tasks: 321 up-to-date
> Configuration cache entry discarded with 2539 problems.

So…do we want to put up with seeing that for whole-project builds in exchange for making many other build targets faster by default?

**Comment by msridhar** [comment]
My vote is to put it on by default, at least for local builds.  In my development workflow I rarely run a full build.  I'd rather not have to always need to remember to add the configuration cache flag.

I'm wondering, could we just disable the configuration cache on CI?  This way, we will have a check that (while incubating at least) the configuration cache is not leading to erroneously-successful builds.  And we may want to update our release instructions to disable the configuration cache when building release artifacts.

**Comment by liblit** [comment]
> My vote is to put it on by default, at least for local builds.

Sounds good to me! I'll make that happen.

> In my development workflow I rarely run a full build. I'd rather not have to always need to remember to add the configuration cache flag.

Right, that's exactly the scenario I had in mind. I imagined you tediously adding `--configuation-cache` to each command line or IntelliJ IDEA run configuration, and thought that's probably more annoying than the warning messages.

> I'm wondering, could we just disable the configuration cache on CI?

I suppose we could. I see pros and cons to doing this. Pro: knowing that our builds are all hermetic even if there are defects in the caching logic. Con: no longer testing builds in a way that matches how you or me or other users would be running them.

>  And we may want to update our release instructions to disable the configuration cache when building release artifacts.

Do we currently disable the build cache when building release artifacts? It's easier for me to imagine how we could get a non-hermetic, irreproducible build due to a malfunctioning build cache rather than a malfunctioning configuration cache.

**Comment by msridhar** [comment]
> > I'm wondering, could we just disable the configuration cache on CI?
> 
> I suppose we could. I see pros and cons to doing this. Pro: knowing that our builds are all hermetic even if there are defects in the caching logic. Con: no longer testing builds in a way that matches how you or me or other users would be running them.

Since we mostly run `./gradlew build` on CI maybe the config cache won't kick in there anyway?  For me, it depends on how concerned we are about bugs in the Gradle config cache support itself.  Since it's incubating it may be good to have a sanity check in case config cache causes a build inconsistency.

> 
> > And we may want to update our release instructions to disable the configuration cache when building release artifacts.
> 
> Do we currently disable the build cache when building release artifacts? It's easier for me to imagine how we could get a non-hermetic, irreproducible build due to a malfunctioning build cache rather than a malfunctioning configuration cache.

We don't disable the build cache right now.  Maybe we should?  I guess it's been fine thus far.  So I guess we can leave it and probably leave the config cache on also.
